portfolio-project-matching-app/backend/dao.js

```diff
@@ -410,6 +430,56 @@ const deleteAssociation = async (coll, id1, id2) => {
     }
 }
 
+const deleteDocAndAssociations = async (coll, id) => {
+    /*
+    DESCRIPTION:    deletes document with provided document ID from provided
+                    collection. Also deletes any association documents e.g.,
+                    'projects_users', 'users_technologies', etc
+
+    INPUT:          coll (string) : name of Firebase collection where the
+                    document being updated is located
+
+                    id (string) : document ID of document being deleted
+
+    RETURN:         NA
+    */
+    // get document snapshot for invalid input handling
+    const snap = await getDocSnapshotById(coll, id);
+    // handle case where id does not exist in provided Firebase collection
+    if (!snap.exists()) {
+        console.log(`invalid ${coll} document: '${id}' does not exist`);
+        return -1;
+    }
+    // handle case where inputs are valid
+    else {
+        // delete associations using helper object
+        for (const map of deleteAssociationsHelper[coll]) {
+            const querySnap = await readQuerySnapshotById(map.coll, map.field, id);
+            for (const doc of querySnap.docs) {
+                await deleteDoc(map.coll, doc.id);
+            }
+        }
+        // delete main doc
+        const docRef = await deleteDocById(coll, id);
+        console.log(`Deleted ${coll} document with id: '${docRef.id}'`);
+        return docRef;
+    }
+}
+
+// helpers, don't export
+
+const deleteAssociationsHelper = {
+    'projects': [
+        {coll: 'projects_users', field: 'project_id'},
+        {coll: 'projects_technologies', field: 'project_id'}
+    ],
+    'users': [
+        {coll: 'projects_users', field: 'user_id'},
+        {coll: 'projects_technologies', field: 'user_id'}
+    ],
+    'technologies': []
+};
```

Comment:
This helper object removed a ton of repetitive looping based on which collection the "original" document resides in.